Secure element key creation foundation

gilles-peskine-arm opened the pull request:
This PR lays some foundations for creating a key in a secure element. It's work in progress, not yet usable.

What it does so far:

* Support for import, export and destroy. Any other function on a secure element key is likely to cause a crash.
* The metadata about SE keys is stored persistently. Not all failure cases are handled.
* Under the hood, there's a mechanism that functions that operate on keys can use to find out which SE driver handles a key if any (implemented for `export`). Every such function will need to have this mechanism. I don't think the existing code is the final word on this.
* The core stores a persistent data buffer on behalf of the driver. The content of the buffer is fully up to the driver. It can be, for example, a slot usage map.
* Slot allocation is fully up to the driver. The core stores a 64-bit “slot number” for each key.

This PR has reached the goals I wanted to reach for a foundation:

* Provide enough working code to pass a smoke test (I chose import-export-destroy).
* Implement enough of the API to see if the information is in the right place where it's needed. I haven't yet implemented the two-phase commit that is necessary to handle power failures during key creation or destruction, but I'm confident that it can be implemented with the current API.

I think the path towards getting this merged is to implement just enough that doing something that isn't implemented yet on a key that is in a secure element would return an error instead of crashing. Implementing more secure element access hooks at this point would be premature: first we should beef up testing.

History:
1. Original https://github.com/gilles-peskine-arm/mbed-crypto/tree/psa-se_driver-create_key-2
    * Secure element drivers must declare how many “slots” they support, and the core maintains a bitmap indicating which slots are free.
    * Secure element drivers can enforce restrictions on which slot numbers can store which types of keys.
2. 2019-07-12
    * The core stores a persistent data buffer on behalf of the driver. The content of the buffer is fully up to the driver. It can be, for example, a slot usage map.
    * Slot allocation is fully up to the driver. The core stores a 64-bit “slot number” for each key.

AndrzejKurek commented on library/psa_crypto_se.c:
```diff
@@ -32,13 +32,53 @@
+    psa_key_lifetime_t lifetime )
+{
+    size_t i;
+    if( lifetime == 0 )
```
I have the same remark as in a different PR - I would suggest swapping zero with a defined constant to signal what does this value mean.

gilles-peskine-arm replied:
Here 0 means no entry. I'll add a comment to explain.

AndrzejKurek commented on library/psa_crypto_se.c:
```diff
@@ -83,4 +123,10 @@ void psa_unregister_all_se_drivers( void )
 
+
+
+/****************************************************************/
```
What is the reason behind this comment?

gilles-peskine-arm replied:
Makes it visible that this finishes the "Driver registration" section.

AndrzejKurek replied:
It's the end of the file anyway and nothing is after this section, that's why I ask. It seems redundant to me.

gilles-peskine-arm replied:
No changes planned here.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -1390,12 +1402,25 @@ static psa_status_t psa_finish_key_creation( psa_key_slot_t *slot )
         return;
+
+#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
+    /* TOnogrepDO: If the key has already been created in the secure
```
This shouldn't be here, at least not in such strange form.

gilles-peskine-arm replied:
That's why it's a TODO. Spelled this way so that `check-files` doesn't tell me something I already know.

Patater replied:
Why not fix check-files in a separate commit instead?

Also, please add a reference in this code comment to a GitHub issue for this TODO, if it's important enough to track fixing.

gilles-peskine-arm replied:
It was a conscious decision to reject `TODO` in `check-files.sh`, so I'm respecting it. Changing our criteria for acceptable code is out of scope of this PR.

gilles-peskine-arm replied:
TOnogrepDO will stay in this PR. When https://github.com/ARMmbed/mbedtls/pull/2730 makes its way to this branch, we can use TODO.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -1528,12 +1565,32 @@ psa_status_t psa_import_key( const psa_key_attributes_t *attributes,
+            slot->data.se.slot_number,
+            slot->lifetime, slot->type, slot->policy.alg, slot->policy.usage,
+            data, data_length );
+        /* TOnogrepDO: psa_check_key_slot_attributes? */
```
This shouldn't be here, at least not in such strange form.

gilles-peskine-arm replied:
Will be resolved in a follow-up PR.

AndrzejKurek commented on include/psa/crypto_se_driver.h:
```diff
@@ -677,6 +677,49 @@ typedef struct {
+ * \param before            The end of the range.
+ *                          It is not included in the search.
+ * \param[out] found        On success, a free slot number such that
+ *                          `from <= found < before`.
```
What is the reason behind such rule? It will be hard to remember, since it's inconsistent even within this one function call.

gilles-peskine-arm replied:
I don't understand. Why is it so strange that the slot is within an interval, since that's the whole point of the function? What inconsistency are you talking about?

AndrzejKurek replied:
I'm talking about having the beginning of the range inclusive, but the end - not.

gilles-peskine-arm replied:
That's the usual way to do ranges, so that end - start = length. What would you expect?

AndrzejKurek replied:
I would expect having the same rules for both sides of the range. I'm not acquainted with other crypto APIs though, so I won't press on changing it.

gilles-peskine-arm replied:
It's not about crypto APIs, it's about array slices. Most array slicing APIs take (start_inclusive, number_of_items) or (start_inclusive, end_exclusive) as parameters, with end_exclusive = start_inclusive + number_of_items. If you use start_inclusive and end_inclusive, a lot of operations need a +1 or -1, which is awkward and error-prone.

gilles-peskine-arm replied:
But maybe this would look more familiar with a length parameter instead of an end parameter. I'm considering changing it.

gilles-peskine-arm replied:
Moot because this interface has been completely redesigned.

AndrzejKurek commented on include/psa/crypto_se_driver.h:
```diff
@@ -677,6 +677,49 @@ typedef struct {
+ *         Success. \c *found contains a slot number which is between
+ *         \p from and \p before-1 inclusive and which is currently free.
+ * \retval #PSA_ERROR_INSUFFICIENT_STORAGE
+ *         All slots in the indicated range are occupied.
```
Minor suggestion: `indicated` -> `requested`

gilles-peskine-arm replied:
Moot because this chunk of text is no longer present.

AndrzejKurek commented on library/psa_crypto_se.c:
```diff
@@ -75,6 +90,123 @@ const psa_drv_se_t *psa_get_se_driver( psa_key_lifetime_t lifetime )
+        return( PSA_ERROR_INSUFFICIENT_MEMORY );
+    drv->slot_usage->size = drv->methods->key_management->slot_count;
+
+    /* TOnogrepDO: load */
```
Leftover comment

gilles-peskine-arm replied:
No, it's still to do.

Patater replied:
Should this PR be blocked on loading not being done yet?

gilles-peskine-arm replied:
It shouldn't be merged as is because it includes a feature that's on by default (SE support) but leads to undefined behavior if used according to the documentation. I don't have precise criteria in mind for when this PR will be mergeable: ideally it would do everything but I also want it to be usable ASAP.

Patater replied:
I'd recommend not making this PR do everything as a precondition to merging.

Let's try to pick a safe subset to enable by default, and implement that bit. Or, make this off by default until we are comfortable enough with the feature to turn it on by default. The latter approach is quite common and widely known in the industry as a "feature toggle".

gilles-peskine-arm replied:
I'm not planning to remove TOnogrepDO's from this PR in general. This particular one has actually been done.

AndrzejKurek commented on library/psa_crypto_se.c:
```diff
@@ -28,18 +28,33 @@
+#define mbedtls_free   free
+#endif
+
+
```
Triple whitespace

gilles-peskine-arm replied:
What's triple whitespace and why is it a problem?

AndrzejKurek replied:
_Pars pro toto_ - Three consecutive newline whitespaces. I'm quite certain I haven't seen these in any of the previous MbedTLS files, they seem a bit spacious. Not sure if we have any rule against it though. Pointing it out just in case it's an oversight.

gilles-peskine-arm replied:
No changes planned.

Patater commented on library/psa_crypto_se.c:
```diff
@@ -83,4 +123,10 @@ void psa_unregister_all_se_drivers( void )
+
+/****************************************************************/
+/* The end */
+/****************************************************************/
```
Please add ASCII art of a man taking a bow. It should fit very well with this "The end" comment.

gilles-peskine-arm replied:
ASCII art request denied.

Patater replied:
:-(

Patater commented on library/psa_crypto_se.c:
```diff
@@ -75,6 +90,123 @@ const psa_drv_se_t *psa_get_se_driver( psa_key_lifetime_t lifetime )
+struct psa_drv_se_slot_usage_s
+{
+    size_t size; /** Number of elements of \c vector */
+    bv_element_t vector[1]; /* Should be vector[] but we don't allow C99 */
```
We do allow C99. Use vector[]. CI won't break.

gilles-peskine-arm replied:
Unfortunately, no, we don't allow C99 (I checked). We only allow the C99 features that Visual Studio 2010 supports. I don't think this includes `vector[]`: there's an error on line 205 in https://jenkins-internal.mbed.com/job/mbedtls-release-new/304/execution/node/319/log/
```
202     typedef struct
203     {
204         int foo;
205         int bar[];
206     } incomplete_t;
```

Patater replied:
Mistake on us to be using Visual Studio 2010 in our CI still, then.

gilles-peskine-arm replied:
Moot because this chunk of code is no longer present.

Patater commented on library/psa_crypto_se.c:
```diff
@@ -75,6 +90,123 @@ const psa_drv_se_t *psa_get_se_driver( psa_key_lifetime_t lifetime )
+    bv_element_t vector[1]; /* Should be vector[] but we don't allow C99 */
+};
+
+static psa_key_slot_number_t su_size(
```
Secure Ulement? Slot Usage? Slow Umbrella? Use more descriptive name please, to improve readability.

gilles-peskine-arm replied:
Moot because this chunk of code is no longer present.

Patater commented on library/psa_crypto_se.c:
```diff
@@ -75,6 +90,123 @@ const psa_drv_se_t *psa_get_se_driver( psa_key_lifetime_t lifetime )
+        *elt = *elt & ~mask;
+}
+
+/* Callback function for drivers */
```
For drivers to implement? for drivers to use? to copy paste? What should drivers do with this?

gilles-peskine-arm replied:
Moot because this chunk of code is no longer present.

Patater commented on library/psa_crypto_se.c:
```diff
@@ -28,18 +28,33 @@
 {
     psa_key_lifetime_t lifetime;
     const psa_drv_se_t *methods;
+    psa_drv_se_slot_usage_t *slot_usage;
```
Just an observation, looking for a check on my understanding:

It looks like we are currently requiring separate lifetimes for slots that have different capabilities. Is this correct? Is there no other way to correlate what a slot is capable of doing from lifetime?

This seems to require that for raw data import/export slots used for e.g. storing certificates would have a separate lifetime from slots used for ecdsa keys.

Do we also require a new lifetime value for slots that have different hardware-enforced policies? For instance, some slots may be configured as read-only. We don't want to allocate a slot that's read-only for import-key use. We want the slots returned by `psa_drv_cb_find_free_slot()` to be fit for purpose and OK by the desired usage policy.

gilles-peskine-arm replied:
Implementers have some freedom in choosing lifetimes, but the intent is that they characterize the lifetime of a key, not what type of key it is. So slots that can store different types of keys should still have the same lifetime, but erasable and non-erasable slots in the same secure element should have different lifetimes.

Driver can have an `allocate_key` method precisely so that they can pick a slot that's suitable for a given key's parameters (lifetime, type, size, policy, …).

Patater replied:
Thanks for the explanation. This makes sense now that I understand policy is part of attributes, so that we don't need a new `lifetime` for each policy/usage required.

Patater commented on include/psa/crypto_se_driver.h:
```diff
@@ -685,6 +728,32 @@ typedef struct {
+typedef psa_status_t (*psa_drv_se_allocate_key_t)(
+    const psa_key_attributes_t *attributes,
+    const psa_drv_se_slot_usage_t *slot_usage,
+    psa_key_slot_number_t *key_slot);
```
Should the policy also be provided here, so that the driver can avoid allocating a key that would be unsuitable for the desired use case? It'd stink to be an application and have to call allocate a bunch of times in a loop until you got allocated a key that worked with your desired usage policy.

gilles-peskine-arm replied:
The policy _is_ provided. It's part of `attributes`.

Patater replied:
Excellent.

Patater commented on include/psa/crypto_se_driver.h:
```diff
@@ -980,12 +980,12 @@ typedef struct {
-    psa_drv_se_aead_t aead;
-    psa_drv_se_asymmetric_t asymmetric;
-    psa_drv_se_key_derivation_t derivation;
+    const psa_drv_se_key_management_t *key_management;
```
Consider also making the pointer itself const, if possible. Without that, you can't get any of that sweet armclang LTO action.

gilles-peskine-arm replied:
Isn't it enough to make the whole structure `const`?

Patater replied:
Needs testing.

gilles-peskine-arm replied:
I'm not planning to change anything here in this PR. It's a more general issue for the driver interface anyway. If you think a change is needed, please file it separately in https://github.com/ARMmbed/psa-crypto/issues.

shelib01 commented on include/psa/crypto_se_driver.h:
```diff
@@ -726,7 +726,7 @@ typedef psa_status_t (*psa_drv_se_import_key_t)(psa_key_slot_number_t key_slot,
+typedef psa_status_t (*psa_drv_se_destroy_key_t)(psa_key_slot_number_t key_slot);
 
 /**
  * \brief A function that exports a secure element key in binary format
```
what about generate key ?

shelib01 commented on include/psa/crypto_se_driver.h:
```diff
@@ -726,7 +726,7 @@ typedef psa_status_t (*psa_drv_se_import_key_t)(psa_key_slot_number_t key_slot,
  *         The slot's content, if any, has been erased.
  */
-typedef psa_status_t (*psa_drv_se_destroy_key_t)(psa_key_slot_number_t key);
+typedef psa_status_t (*psa_drv_se_destroy_key_t)(psa_key_slot_number_t key_slot);
```
what about generate key ?

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -1455,23 +1559,44 @@ psa_status_t psa_import_key( const psa_key_attributes_t *attributes,
+            status = PSA_ERROR_NOT_SUPPORTED;
+            goto exit;
+        }
+        status = drv->key_management->p_import(
```
Bug found by https://github.com/ARMmbed/mbed-crypto/pull/164/commits/4f49e61a7908ecde0008bb387c0c91af845b3e43 : 
Lack of error checking

gilles-peskine-arm replied:
What error checking specifically? Since we already have a test, whatever that bug is, I propose to fix it in a follow-up PR.

Patater commented:
### TODO before merge

- Complete code review
- Use feature toggle approach
  - Disabled by default until we are confident in the feature being safe to use
  - Don't break non-SE-use-case when SE feature is disabled
  - Don't knowingly break non-SE-use-case when SE feature is enabled
- all.sh
  - Test full configuration without SE
  - Test baremetal configuration without SE
  - Do a single test with SE enabled (using the minimal tests we have in this PR today)

### TODO after merge

- all.sh
  - Enable in Full configs
  - Enable in default
  - New test with SE disabled
- To be safe for production:
  - Need power cycling tests 
    - Can do mock power cycle tests
    - Between and during operations
  - Need to check every function that does something with a key
    - Avoid crashes if they say, try to access key material 
    - Adding hooks to call driver and testing those, including error handling
    - Fix bugs in the code added by this PR due to lack of error handling
  - Storage failure testing
    - When storage fails, we handle it OK

AndrzejKurek commented:
Mbed-os problems encountered during atecc608-a example update:

- mbed-os\components\TARGET_PSA\services\attestation\COMPONENT_PSA_SRV_IMPL\attest_crypto.c - `psa_open_key` in function `t_cose_crypto_pub_key_sign` - passing argument 2 of 'psa_open_key' makes pointer from integer without a cast.
- mbed-os\features\mbedtls\mbed-crypto\platform\COMPONENT_PSA_SRV_IMPL\psa_crypto_storage.c - `psa_its_get` in function `psa_crypto_storage_load` - too few arguments;

gilles-peskine-arm commented:
@AndrzejKurek Are these errors related to this PR? No code in Mbed OS should already be using secure element drivers.

AndrzejKurek commented:
No, it's related to updating the crypto version in mbed-os - I'm just supplementing Jaeden's todo list, the "after merge" part to be specific.  I'll be re-reviewing this PR now.

gilles-peskine-arm commented:
@AndrzejKurek Updating the crypto version in Mbed OS is completely unrelated to this PR. We can't update Mbed OS to this branch until this branch supports TLS anyway.

gilles-peskine-arm commented:
To make sure that the consequences are not too bad if someone turns on secure element support, I've ensured that at least calling a function that doesn't support SE keys on an SE key will error out and not attempt to access nonexistent fields in the slot object.

gilles-peskine-arm commented:
[Crypto CI](https://jenkins-internal.mbed.com/blue/organizations/jenkins/mbed-crypto-pr/detail/PR-157-head/8/pipeline) with [SE enabled by default](https://github.com/ARMmbed/mbed-crypto/pull/157/commits/105736653ffcfa530fe989b81880acffc2c02441) is good (only ABI check fails, for expected reasons).

gilles-peskine-arm commented:
[Crypto CI](https://jenkins-internal.mbed.com/job/mbed-crypto-pr/job/PR-157-head/9/display/redirect) with SE disabled by default but tested by two all.sh components passed (except ABI check, but that's expected).

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -1704,7 +1737,7 @@ psa_status_t psa_copy_key( psa_key_handle_t source_handle,
 
-    status = psa_get_key_from_slot( source_handle, &source_slot,
+    status = psa_get_transparent_key( source_handle, &source_slot,
                                     PSA_KEY_USAGE_COPY, 0 );
```
Indentation is off.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -5282,7 +5315,7 @@ psa_status_t psa_key_derivation_input_key(
     psa_key_slot_t *slot;
     psa_status_t status;
-    status = psa_get_key_from_slot( handle, &slot,
+    status = psa_get_transparent_key( handle, &slot,
```
Indentation is off.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -5431,7 +5464,7 @@ psa_status_t psa_key_derivation_key_agreement( psa_key_derivation_operation_t *o
     if( ! PSA_ALG_IS_KEY_AGREEMENT( operation->alg ) )
         return( PSA_ERROR_INVALID_ARGUMENT );
-    status = psa_get_key_from_slot( private_key, &slot,
+    status = psa_get_transparent_key( private_key, &slot,
```
Indentation is off.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -5459,7 +5492,7 @@ psa_status_t psa_raw_key_agreement( psa_algorithm_t alg,
         goto exit;
     }
-    status = psa_get_key_from_slot( private_key, &slot,
+    status = psa_get_transparent_key( private_key, &slot,
```
Indentation is off.

Patater commented on library/psa_crypto_se.c:
```diff
@@ -28,23 +28,224 @@
+    psa_status_t storage_status;
+    if( driver->methods->key_management == NULL ||
+        driver->methods->key_management->p_destroy == NULL )
+        return( PSA_ERROR_NOT_PERMITTED );
```
Should this be `PSA_ERROR_NOT_SUPPORTED` instead? The driver doesn't have the function pointer, so it seems the driver doesn't support removal, not that removal fails due to some permissions check.

gilles-peskine-arm replied:
Not being able to destroy a key that you created is iffy. What if the key is compromised and you don't want to use it anymore? Or what if you don't want to use the key anymore but want to keep it secure, and so you want to protect it against a future compromise? So `psa_destroy_key` does not allow the return value `NOT_SUPPORTED`.

The only case where key destruction should be impossible is if there's no way to destroy the key at any level of authorization, for example in a driver for a secure element where all keys are locked. (There may be still be a way to do some kind of factory reset, using nonstandard APIs.) In this case, the caller is authorized to use keys, but not to destroy these keys. Hence `NOT_PERMITTED`.

I'll add a comment to explain, because this is indeed surprising without context.

Patater commented on library/psa_crypto_se.c:
```diff
@@ -28,23 +28,224 @@
+    if( driver->methods->key_management == NULL ||
+        driver->methods->key_management->p_destroy == NULL )
+        return( PSA_ERROR_NOT_PERMITTED );
+    status = driver->methods->key_management->p_destroy(
```
Style: Good job using a cleaner function pointer calling style here. No change requested.

Patater commented on include/psa/crypto_se_driver.h:
```diff
@@ -40,10 +40,85 @@
+     * allowed to modify the persistent data. These functions receive a
+     * writable pointer.
+     */
+    const void *const persistent_data;
```
What would this be used for? We are going through a lot of work to save and restore this, but I don't see the point yet.

gilles-peskine-arm replied:
Oops, that disappeared in some earlier rework. The driver can use it for whatever it wants, but it's geared towards keeping track of which slots are in use. I'm rewriting the description.

Patater commented on library/psa_crypto_slot_management.c:
```diff
@@ -124,15 +124,35 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *p_slot )
-                                       key_data, key_data_length );
+    p_slot->lifetime = psa_get_key_lifetime( &attributes );
+    p_slot->type = psa_get_key_type( &attributes );
+    p_slot->policy = attributes.policy;
```
Could we just store attributes in the slot? Then we could directly `psa_load_persistent_key()` into the `p_slot->attributes`

gilles-peskine-arm replied:
I thought about this, as it would simplify the design. The thing is, the layout of `psa_key_attributes_t` may change without warning and that's just an ABI (not API) change, whereas the layout of storage must not change without providing an upgrade mechanism, otherwise you can't upgrade the code on an existing device. There may be a better way to do it (and it's even more visible in https://github.com/ARMmbed/mbed-crypto/pull/183 which also saves the bit-size and the slot number for SE keys), but that would be going too far for this PR.

Patater commented on library/psa_crypto_slot_management.h:
```diff
@@ -93,6 +122,7 @@ psa_status_t psa_internal_allocate_key_slot( psa_key_handle_t *handle,
 psa_status_t psa_validate_persistent_key_parameters(
     psa_key_lifetime_t lifetime,
     psa_key_file_id_t id,
+    psa_se_drv_table_entry_t **p_drv,
```
Should this be `const psa_se_drv_table_entry_t **p_drv`? I don't see a reason we need to change the driver table it self in this function, although we do want to change the value of the pointer and don't want to make the pointer const.

gilles-peskine-arm replied:
I don't think so: it would require the caller to pass a `const psa_se_drv_table_entry_t **`, and then the caller('s caller) wouldn't be able to change any part of the table entry, specifically the mutable part of the context that the driver is allowed to modify (`driver->context->transient_data`).

I did try to make more things const around driver table entries, but I didn't find a way.

Patater replied:
Implicitly casting to a function that accepts a const is fine. It doesn't have to be actually const for the caller('s caller).

gilles-peskine-arm replied:
What do you mean by “implicitly casting”? I can't find a way to add the `const`s that I'd like to add without adding (explicit) casts, which are a maintenance burden at best and cause undefined behavior due to lying to the compiler at worst.

Patater replied:
I played around with this as well. Looks I was wrong about this. We can't avoid explicit casts in C if we want const correctness here, due to the nested pointer.

Patater commented on library/psa_crypto_storage.c:
```diff
@@ -249,6 +255,7 @@ static psa_status_t psa_crypto_storage_get_data_length(
     uint8_t version[4];
+    uint8_t lifetime[sizeof( psa_key_lifetime_t )];
     uint8_t type[sizeof( psa_key_type_t )];
     uint8_t policy[sizeof( psa_key_policy_t )];
```
Should we store uint8_t attributes[] instead of splitting it up into a separate type and policy?

gilles-peskine-arm replied:
See https://github.com/ARMmbed/mbed-crypto/pull/157#discussion_r306865039

Patater commented on tests/scripts/list-identifiers.sh:
```diff
@@ -41,7 +41,7 @@ rm -f identifiers
 grep '^[^ /#{]' $HEADERS | \
     sed -e 's/^[^:]*://' | \
-    egrep -v '^(extern "C"|(typedef )?(struct|enum)( {)?$|};?$)' \
+    egrep -v '^(extern "C"|(typedef )?(struct|union|enum)( {)?$|};?$)' \
```
Dear Gilles,

Thanks for letting Mr. Check Names know about this essential keyword.

Sincerely,
Potato man

Patater commented on tests/suites/test_suite_psa_crypto_se_driver_hal.data:
```diff
@@ -26,3 +26,72 @@ register_twice:3
+SE key import-export, check after restart (slot 3)
+key_creation_import_export:3:1
+
+Key creation smoke test: AES-CTR
```
Please avoid smoking and say what the test does. The indirection through some new "smoke" concept doesn't help clarity, although it is cute.

If the "smoke" concept includes many other tests, please split the test up into smaller ones that can have clear names. One big long name saying all the tests that "smoke" does wouldn't be helpful.

Alternatively, @AndrzejKurek and @dgreen-arm may replace all this smoke testing as temporary and replace it with mocks that check for expected status codes instead of just "valid" status codes. In which case, do nothing!

gilles-peskine-arm replied:
If you have a better name, please suggest. The goal of this test is to call lots of functions on the key and check that the system doesn't do anything crazy. I think this test will still have value even when there are tests for the actual intended behavior: it's easy to check that it's exhaustive (in the sense of calling all API functions that take a key as input).

Patater replied:
Let's go with the alternative and use full-coverage mock-based with better names in a follow up PR. Smoking allowed in this PR.

Patater commented on tests/suites/test_suite_psa_crypto_se_driver_hal.function:
```diff
@@ -3,10 +3,328 @@
+    for( id = 0; id <= MAX_KEY_ID_FOR_TEST; id++ )
+        psa_destroy_persistent_key( id );
+    /* Purge the transaction file. */
+    psa_crypto_stop_transaction( );
```
`psa_purge_storage()` seems to depend on `MBEDTLS_PSA_CRYPTO_SE_C` and `PSA_CRYPTO_STORAGE_HAS_TRANSACTIONS`. Assuming we don't need ifdefs here because this entire file depends on `MBEDTLS_PSA_CRYPTO_SE_C`.

gilles-peskine-arm replied:
`MBEDTLS_PSA_CRYPTO_SE_C` implies `PSA_CRYPTO_STORAGE_HAS_TRANSACTIONS`.

AndrzejKurek commented on docs/architecture/mbed-crypto-storage-specification.md:
```diff
@@ -193,3 +193,92 @@ The layout of a key file is:
+
+#### Transaction types on TBD
+
+* 0x0001: key creation. The following locations may or may not contain data about the key that is being created:
```
What do you mean by `may or may not contain`? If these operations cannot be completed atomically, how about saving information about the stage that was finished?

gilles-peskine-arm replied:
“May or may not” means that you don't know which is the case.

Storing information about the stage that was finished would be costly, and would only be useful if the transaction is interrupted by a power failure (rare case) and only if some more efficient recovery is possible (maybe in some circumstances, but usually not).

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -1491,9 +1501,14 @@ static psa_status_t psa_finish_key_creation(
-            mbedtls_platform_zeroize( buffer, buffer_size );
-        mbedtls_free( buffer );
+#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
+        if( driver == NULL )
```
Looking at all of the hooks and checks for null drivers (here and in other functions) makes me wonder - why not introduce a table of pointers to functions (like in `cipher`) that would be populated on key registration/creation and then, instead of having checks in each function in runtime, a function dispatching mechanism would take care of it?
It probably would decrease complexity at the cost of code size though.

gilles-peskine-arm replied:
I'd like a mechanism that doesn't require so much boilerplate on each call, but how to do it? Pretty much each method takes a different set of parameters. A variadic macro might help. But that's out of scope here, it's a job for https://github.com/ARMmbed/mbed-crypto/pull/174.

AndrzejKurek commented on library/psa_crypto_storage.c:
```diff
@@ -399,6 +405,60 @@ psa_status_t psa_load_persistent_key( psa_key_file_id_t key,
+
+psa_crypto_transaction_t psa_crypto_transaction;
+
+psa_status_t psa_crypto_save_transaction( void )
```
Is it possible to provide an alternative implementation of transaction storage, for example for targets that will have a custom persistent memory mechanism (cloud/EEPROM/etc)?

gilles-peskine-arm replied:
I guess it is, but I haven't seen any request for it. If you want an alternative mechanism, you'd only need to replace the load/save functions, so the architecture of the code supports it. Making an alternative mechanism pluggable without editing the source code is way out of scope here, especially with no concrete use case.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -950,7 +950,20 @@ psa_status_t psa_destroy_key( psa_key_handle_t handle )
+        status = psa_crypto_save_transaction( );
+        if( status != PSA_SUCCESS )
+        {
+            /* TOnogrepDO: destroy what can be destroyed anyway */
```
`psa_crypto_stop_transaction( );`?

gilles-peskine-arm replied:
Good question. In practice this currently wouldn't make any difference: if `save` failed, the file wasn't created. So the only thing `stop` would do is to zero out the transaction object in memory, which is good hygiene but wouldn't affect the behavior of the code. Should `save` do this on failure? It's the right thing in all cases currently because `save` is only ever used once per transaction. But if `save` was used in the middle of a transaction, the caller might need to read back some of that information on a failure.

gilles-peskine-arm replied:
I went for your suggestion, which is the one that's conceptually simplest.

AndrzejKurek commented on library/psa_crypto.c:
```diff
@@ -1382,15 +1407,24 @@ static psa_status_t psa_start_key_creation(
+    psa_crypto_transaction.key.slot = slot->data.se.slot_number;
+    psa_crypto_transaction.key.id = slot->persistent_storage_id;
+    status = psa_crypto_save_transaction( );
+    if( status != PSA_SUCCESS )
```
`psa_crypto_stop_transaction( );`?

gilles-peskine-arm replied:
Same as https://github.com/ARMmbed/mbed-crypto/pull/157#discussion_r306888511

gilles-peskine-arm commented on include/psa/crypto_se_driver.h:
```diff
@@ -109,15 +186,15 @@ typedef psa_status_t (*psa_drv_se_mac_update_t)(void *p_context,
  * operation by comparing the resulting MAC against a provided value
  *
- * \param[in,out] p_context A hardware-specific structure for the previously
+ * \param[in,out] op_context    A hardware-specific structure for the previously
```
I'll do a pass of aligning, because that's the general rule in these files. But in the medium term I think we should stop aligning because it has a high maintenance cost and it hurts readability. It causes all the text to be compressed into a very narrow column.

gilles-peskine-arm commented on include/psa/crypto_se_driver.h:
```diff
@@ -75,28 +151,29 @@ typedef uint32_t psa_key_slot_number_t; // Change this to psa_key_slot_t after p
  */
-typedef psa_status_t (*psa_drv_se_mac_setup_t)(void *p_context,
+typedef psa_status_t (*psa_drv_se_mac_setup_t)(psa_drv_se_context_t *drv_context,
+                                               void *op_context,
```
`drv_context` is the global driver context, passed most methods. `op_context` is the per-operation context, passed to methods in a multipart operation. The description of `op_context` at least could use some improvement, and the design might be improved by a few tweaks. This is out of scope of this PR, tracking internally at https://github.com/ARMmbed/psa-crypto/issues/230 .

gilles-peskine-arm commented on include/psa/crypto_types.h:
```diff
@@ -133,6 +133,119 @@ typedef uint32_t psa_key_usage_t;
+ *    You may call psa_reset_key_attributes(), although this is optional
+ *    with the workflow presented here because the attributes currently
+ *    defined in this specification do not require any additional resources
+ *    beyond the structure itself.
```
I really don't understand where you're coming from. Why on earth would `psa_reset_key_attributes` have any effect on a secure element?

Patater replied:
I guess if you think of this as affecting the attributes only, there is no reason for it to affect a SE.

My angle was if attributes allocate memory internally, and have some pointer in them, you'd want that freed with reset. A SE slot is like a pointer, and a SE slot in the SE itself is like the allocated memory, so one may want the slot freed in SE as well via reset.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -925,17 +972,53 @@ psa_status_t psa_destroy_key( psa_key_handle_t handle )
+            return( status );
+        }
+    }
+#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
```
I'll add a comment, but maybe part of the confusion is that you seem to expect that `psa_crypto_save_transaction` and `psa_crypto_stop_transaction` are somehow redundant. Actually `save` starts the transaction and `stop` marks it as finished. Do you think the function names should be changed?

Patater replied:
Doesn't stop remove the transaction from storage? If so, then I'm not sure how the transaction is committed before it disappears.

Given your prose description, we should use terms that match as that's probably the more intuitive way to think about it. This will match the shape of the rest of the API as well. psa_crypto_start_transaction() and psa_crypto_finish_transaction().

gilles-peskine-arm replied:
There's no function that I'd really want to call `start`. The first function `prepare`, which does not save anything to storage. Should it be called `start`, to be paired with `stop`? But `stop` _does_ access storage. I find `finish` less descriptive than `stop` for what it does, because I'd expect `finish` to do some kind of committing, but `stop` doesn't, it just removes the information about a transaction in progress.

Patater replied:
OK, I understand this mechanism now. The "committing" is happening in `psa_destroy_se_key()` in this case. We are saying "I'm going to do this." and the we do it, and then we say "I finished doing this."

In crypto init, if we have a pending transaction, we have enough information to complete the pending transaction (or check that it was done already). I'm not sure when we'd choose to roll back the transaction, though, nor what operation we'd perform to do the rollback (like adding the key back to the key slot for example, since the SE never completed removal, maybe).

athoelke replied:
Rolling back a key destruction is not viable - we cannot determine if the key material is or is not present in the SE and recover it if not. But as this API is destructive, it is easy to record enough information to retry/complete the key destruction.

For key creation/import/derivation, adopting a rolling-back approach means that the transaction does not need to retain any key material, only the key meta-data. On restart, an incomplete transaction can be rolled back by deleting any maybe-created key material. Adding an extra save of the transaction entry after the SE has created the material to record that fact before making updates to the driver's persistent data and the core's key meta-data records would enable a roll-forward as the secrets have been populated in the SE.

athoelke replied:
... but that adds an extra transaction update to benefit a rare situation (system fail during persistent SE key creation).

Patater replied:
> Rolling back a key destruction is not viable

it depends on the order of operations, doesn't it? If we start the transaction, remove the information from core, and then fail before removing the information from the SE, we could put the information back into the core given information in the transaction. If we really can't determine if the key is partially removed or still present in the SE, then I agree, even with this order of operations, we can't roll back the destruction as we need a way to validate that the key is still in the SE (as we could have failed/lost power after SE removal, but before transaction completion).

Not necessary for this PR, but we should write down what our policies should be (at least for our implementation, not necessarily anything at the spec level) for when we roll-back and when we automatically retry a transaction (if ever). We then can test against our written policy with mocks and fault injection to ensure we adhere to our policy.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -1303,11 +1423,14 @@ static psa_status_t psa_set_key_policy_internal(
     psa_status_t status;
     psa_key_slot_t *slot;
 
+    *p_drv = NULL;
```
We do need this because `psa_validate_persistent_key_parameters` is called conditionally.

It is currently the case that `if( p_drv != NULL ) *p_drv = NULL;` in `psa_validate_persistent_key_parameters` will never trigger, but those ~4 bytes of code are worth it to make the function more robust. Uninitialized pointers can be nasty.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -1108,11 +1204,33 @@ static psa_status_t psa_internal_export_key( const psa_key_slot_t *slot,
+                   drv->key_management->p_export );
+        if( method == NULL )
+            return( PSA_ERROR_NOT_SUPPORTED );
+        return( ( *method )( drv_context,
```
I actually prefer explicit dereference, but it isn't worth arguing.

Patater replied:
The API is for others to use, not just yourself. Do what's intuitive and expected by experienced driver authors.

gilles-peskine-arm replied:
This is internal code, not API design, and in any case I don't see why one way or the other would be more expected by experienced driver authors, given that both styles exist in the wild. But as I've said, it's not worth arguing.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -1338,47 +1484,82 @@ static psa_status_t psa_start_key_creation(
+            psa_destroy_persistent_key( slot->persistent_storage_id );
+            return( status );
+        }
+        status = psa_crypto_stop_transaction( );
```
As in https://github.com/ARMmbed/mbed-crypto/pull/157#discussion_r307196730, I think this is obvious: we've finished doing what the transaction is about, so we stop it. What information would you like to see in a comment?

Patater replied:
I was reading stop as "cancel".

Patater replied:
Maybe better to name it "complete" or "finish"?

gilles-peskine-arm commented on library/psa_crypto_storage.h:
```diff
@@ -200,8 +188,169 @@ psa_status_t psa_parse_key_data_from_storage( const uint8_t *storage_data,
+ *    psa_crypto_save_transaction() again when each stage is reached.
+ * -# When the transaction is over, whether it has been committed or aborted,
+ *    call psa_crypto_stop_transaction() to remove the transaction data in
+ *    storage and in memory.
```
I think you have a specific transaction model in mind that works by preparing the new data without touching the old data and, as the last step, atomically switching to the new data (“commit”). But that's not how transactions work here. There's no way to atomically switch to the new data. Any distinction between “abort” and “commit” is in the eyes of the caller. As far as the transaction storage subsystem is concerned, there's just stopping.

Patater replied:
How does anything happen? I see transaction put into storage, nobody using it, and then removed. This has no effect. How do we expect this to work? It may be that we make it work in a follow up PR, but I am curious how we'd want this to work.

I'd expect core to read the transaction (if any) and commit/apply it atomically with other data it is storing, but I don't see any signs of that intent.

athoelke replied:
We probably need a bit more commentary on the transaction model in use.

Because the underlying storage provides single-item atomic replacement, it would be possible to build a multi-item atomic update on top with a suitably designed heirarchy of stored objects. But this measn the the core needs to maintain a directory of data uids which it replaces when running a transaction. Although this is more flexible and permits complex changes to be built into a transaction - we don't actually need this facility.

The model Gilles has built is more like a single-item roll-forward journal, or perhaps even better described as a persistent state machine. The transaction object is just the journal entry, and it records all the data that is needed for the distributed update required to either finish the key operation or roll it back. It also is not generic - it stores operation specific data in order for operation-specific completion/rollback code to handle the situation where the transaction is found on restart.

So for each operation that requires multiple operations on persistent storage, we define a state machine where the current state is stored in the "transaction object". The stored transaction state is enough to permit either a "roll-forward" or a "roll-back" to the key store state - and provide the atomicity required for key store changes.  The recovery mode (rollback/rollfoward), can change for a operation depending on the state-machine state, e.g. when creating a key it will be roll-back until we know the SE has created/stored the key material successfully, the transaction object after that can record a state that allows roll-forward as well as roll-back.

Note that rolling back might fail in some circumstances, as this can require one or more changes to persistent data to accomplish. If it fails, the system could operate by acting "as if the rollback had completed" and retaining the transaction object so that this is remembered. However, that will block any key store operation which would use a new transaction.

Maybe some good verbs for the API would be 'start', 'update' and 'finish', (or 'begin' and 'end') with regards to operations on the persistent transaction state. 'commit', 'abort' and 'stop' may fit a classic transaction model (thinking databases here), but each key store operation that uses this transaction model has a different 'commit point' - when the stored transaction state changes from "must roll back" to "can roll forward"

gilles-peskine-arm replied:
@Patater The storage subsystem provides primitives to store the transaction state. It does not perform the transaction.

The core only ever needs to read back the transaction data if the transaction is interrupted by a reset. (At least so far, this is not intrinsic in the design.)

gilles-peskine-arm commented on library/psa_crypto_storage.h:
```diff
@@ -88,22 +83,19 @@ int psa_is_key_present_in_storage( const psa_key_file_id_t key );
+ * \param[in] attributes    The attributes of the key to save.
+ *                          The key identifier field in the attributes
+ *                          determines the key's location.
+ * \param[in] data          Buffer containing the key data.
```
As far as the storage subsystem is concerned, it's just data. I've actively avoided having it care what this data is. It is somewhat inconvenient for the overall design (https://github.com/gilles-peskine-arm/mbed-crypto/commit/a0081bb3a0a7a423f3f6b6576227a013bf7d714f (part of https://github.com/ARMmbed/mbed-crypto/pull/183) is not very nice), but I'll leave that job for a future PR.

Patater replied:
We should call it data instead of key data, then.

gilles-peskine-arm replied:
“Key data” is ambiguous, which actually works here. In the PSA code and documentation, we normally use “key material” for the representation of the key itself as opposed to its metadata.

Patater replied:
I did not pick up on this distinction between "key data" and "key material", believing that "key data" meant key material, if that was the intent. This would be more readable as just "data".

gilles-peskine-arm commented:
@Patater @AndrzejKurek Thanks for the reviews. I believe I've addressed all your comments except for a few where I'm still waiting for further feedback.

gilles-peskine-arm commented on library/psa_crypto_storage.h:
```diff
@@ -200,8 +188,180 @@ psa_status_t psa_parse_key_data_from_storage( const uint8_t *storage_data,
+ * If the system crashes while a transaction is in progress, psa_crypto_init()
+ * calls psa_crypto_load_transaction() and takes care of completing or
+ * rewinding the transaction. This is done in psa_crypto_recover_transaction()
+ * in psa_crypto.c. If you add a new type of transactions, be
```
Fixed by amending the commit.
